### ISSUE TITLE
Suggesting adding capturable property to the Adam optimizer

### ISSUE THREAD

**Issue by ccffccffcc** [mosaicml/composer#1343]
I was using trainer and Adam optimizer and testing if training can be recovered from a recent checkpoint. It reported 
File "/lib/python3.10/site-packages/torch/optim/adam.py", line 255, in _single_tensor_adam
    assert not step_t.is_cuda, "If capturable=False, state_steps should not be CUDA tensors."
This is because of when initializing the Adam optimizer, capturable is set to False by default. It can be simply solved by adding capturable as a property in the AdamHparams class. (composer.optim.optimizer_hparams.py)

@dataclass
class AdamHparams(OptimizerHparams):
    """Hyperparameters for the :class:`~torch.optim.Adam` optimizer.

    See :class:`~torch.optim.Adam` for documentation.

    Args:
        lr (float, optional): See :class:`~torch.optim.Adam`.
        betas (float, optional): See :class:`~torch.optim.Adam`.
        eps (float, optional): See :class:`~torch.optim.Adam`.
        weight_decay (float, optional): See :class:`~torch.optim.Adam`.
        amsgrad (bool, optional): See :class:`~torch.optim.Adam`.
    """

    lr: float = hp.optional(default=0.001, doc="learning rate")
    betas: List[float] = hp.optional(
        default_factory=lambda: [0.9, 0.999],
        doc="coefficients used for computing running averages of gradient and its square.",
    )
    eps: float = hp.optional(default=1e-8, doc="term for numerical stability")
    weight_decay: float = hp.optional(default=0.0, doc="weight decay (L2 penalty)")
    amsgrad: bool = hp.optional(default=False, doc="use AMSGrad variant")
    capturable: bool = hp.optional(default=False, doc="capturable")

    @property
    def optimizer_object(cls) -> Type[torch.optim.Adam]:
        return torch.optim.Adam

**Comment by ravi-mosaicml** [comment]
Hi there! Thanks for raising this issue. It looks like `capturable` is a new flag that was added in PyTorch 1.12. We can add this flag conditionally on the hparams class.
